### DISCUSSION TITLE
Not able to get out of composing mode

### DISCUSSION THREAD

**Discussion by engeir** [ghostty-org/ghostty#2937] [Feature Requests, Ideas] [answered]
When typing any key that can be used to compose a character (`~`, `` ` ``, `^`, etc.), ghostty is stuck in this mode. See video demonstrating the behavior below. (Also asked in the Discord [here](https://discord.com/channels/1005603569187160125/1315957872663728170)).

I have (the same version of) Ghostty installed by on a Ubuntu machine (where I have the issue) and an Arch Linux machine where I'm not able to reproduce this.

Output of neofetch:

```bash
OS: Ubuntu 24.04.1 LTS x86_64
Host: HP EliteBook 840 G6
Kernel: 6.8.0-49-generic
Uptime: 54 mins
Packages: 4698 (dpkg), 19 (snap)
Shell: zsh 5.9
Resolution: 1920x1080, 2560x1440
WM: bspwm
Terminal: ghostty
CPU: Intel i5-8265U (8) @ 3.900GHz
GPU: Intel WhiskeyLake-U GT2 [UHD Graphics 620]
Memory: 4572MiB / 7748MiB
```

https://github.com/user-attachments/assets/0794f599-d9a3-49f5-99b3-488ecbeaa4c7

**Comment by 00-kat** [comment]
FYI, this commit was already reverted: https://github.com/ghostty-org/ghostty/commit/b19d0d36380af732c61607c7a1669fcb1813290e.

**Reply by engeir** [reply]
I actually still have the same behavior. Just cloned (last commit 9f75d93a551f) and built again, both with `zig build` and `zig build -p $HOME/.local -Doptimize=ReleaseFast` if this matter at all.

EDIT: in the video I'm using the Ghostty version built from that commit (9f75d93a551f).

**Comment by pkielen** [comment]
I have the same issue on current master (6ef757a) with a swedish keyboard. When trying to type `altgr+¨` which should result in `~`.

This is the output from `xev`:
```
KeyPress event, serial 38, synthetic NO, window 0x2e00001,
    root 0x5e3, subw 0x0, time 30322829, (3026,531), root:(3026,600),
    state 0x4010, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes:
    XmbLookupString gives 0 bytes:
    XFilterEvent returns: False

KeyPress event, serial 38, synthetic NO, window 0x2e00001,
    root 0x5e3, subw 0x0, time 30324111, (3026,531), root:(3026,600),
    state 0x4090, keycode 35 (keysym 0xfe53, dead_tilde), same_screen YES,
    XKeysymToKeycode returns keycode: 49
    XLookupString gives 1 bytes: (7e) "~"
    XmbLookupString gives 0 bytes:
    XFilterEvent returns: True

KeyRelease event, serial 38, synthetic NO, window 0x2e00001,
    root 0x5e3, subw 0x0, time 30324247, (3026,531), root:(3026,600),
    state 0x4090, keycode 35 (keysym 0xfe53, dead_tilde), same_screen YES,
    XKeysymToKeycode returns keycode: 49
    XLookupString gives 1 bytes: (7e) "~"
    XFilterEvent returns: False

KeyRelease event, serial 38, synthetic NO, window 0x2e00001,
    root 0x5e3, subw 0x0, time 30325135, (3026,531), root:(3026,600),
    state 0x4090, keycode 108 (keysym 0xfe03, ISO_Level3_Shift), same_screen YES,
    XKeysymToKeycode returns keycode: 92
    XLookupString gives 0 bytes:
    XFilterEvent returns: False
```

Output of neofetch:
```
OS: Ubuntu 24.04.1 LTS x86_64
Host: HP Elite x360 1040 14 inch G10 2-in-1 Notebook PC SBKPF
Kernel: 6.8.0-51-generic
Shell: zsh 5.9
Resolution: 1920x1200, 2560x1440
DE: GNOME 46.0
WM: Mutter
Terminal: ghostty
CPU: 13th Gen Intel i7-1365U (12) @ 5.200GHz
GPU: Intel Raptor Lake-P [Iris Xe Graphics]
Memory: 13940MiB / 31730MiB
```

**Comment by KalleNiemi** [comment]
I got somewhat related issue @ https://github.com/ghostty-org/ghostty/discussions/4552

Issue described here I can not reproduce with swedish keyboard layout, which solves my issue when using alpha version of cosmic-desktop, but persist on new gnome. I'm normally using finnish keyboardlayout. I got bash instead of zsh as you guys are using, maybe the issue is related to that.

Output of neofetch:
```
OS: Arch Linux x86_64 
Host: AB350M Pro4 
Kernel: 6.12.9-arch1-1 
Uptime: 8 mins 
Packages: 1144 (pacman) 
Shell: bash 5.2.37 
Resolution: 3440x1440 
DE: GNOME 47.2 
WM: Mutter 
WM Theme: Adwaita 
Theme: Adwaita [GTK2/3] 
Icons: Papirus-Dark [GTK2/3] 
Terminal: gjs 
CPU: AMD Ryzen 5 5500 (12) @ 4.267GHz 
GPU: NVIDIA GeForce RTX 3060 Lite Hash Rate 
Memory: 2841MiB / 31964MiB 
```

Edit: sorry, if it cannot be reproduced in Arch, my post does not help much :disappointed:

**Comment by emogenet** [comment]
Same thing here

video showing the issue: 

 https://www.youtube.com/watch?v=ub2w0uiuWfA

Also discussed here:

https://github.com/ghostty-org/ghostty/discussions/5056

**Comment by gbbirkisson** [comment]
Same thing here. Hope this gets fixed soon :crossed_fingers: 

I believe this is also the same problem #3164

**Comment by mitchellh** [comment, marked as answer]
Lets move to #3567
